### DISCUSSION TITLE
Help! What does 'ptr' mean in 'from torch_geometric.loader import DataLoader'

### DISCUSSION THREAD

**Discussion by icesky2310** [pyg-team/pytorch_geometric#4332] [Q&A]
Hi, everyone. 
I am just learning the Graph Neural Network. So, I import the class using 'from torch_geometric.loader import DataLoader' and I can see there is an attribute in the dict which named 'ptr'. 
I have looked around the net and went after some useful blogs or references to know what it is. Does anyone please tell me what does that mean? 
`train_loader = DataLoader(train_dataset, batch_size=64)`

One batch in the train_loader  look like this:
`DataBatch(x=[154, 1], edge_index=[2, 124], y=[64], batch=[154], ptr=[65])`

**Comment by rusty1s** [comment]
Thanks for asking. I am very sorry that this is not documented at all (mostly since we only make use of it internally). The `ptr` tensor describes the compressed representation of the `batch` vector (in a similar fashion to the CSR representation of sparse matrices). Each representation can be converted into the other format. For example, if you batch two graphs together (with 4 and 5 nodes, respectively), then `ptr` will denote the offsets between these graphs:
```
[0, 4, 9]
```
In contrast, the `batch` vector will look like:
```
[0, 0, 0, 0, 1, 1, 1, 1, 1]
```
The `ptr` vector allows us to make use of [`segment_reductions`](https://pytorch-scatter.readthedocs.io/en/latest/functions/segment_csr.html), which are a little bit faster than `scatter` operations since we can leverage the fact that values that point to the same index are grouped and sorted in memory.

**Reply by icesky2310** [reply]
Hi, Matthias,
Thanks a lot for your kind help and patient answer which is clear and detailed. 
I have understood it.
Best wishes,
Jianhai Zhang
